File: n8n-ops-ui/src/types/lifecycle.ts
/**
 * Lifecycle stage definitions for Workflow Ops frontend.
 *
 * These stages represent the journey of a workflow through the platform.
 * Note: AUTHORING is not included as it happens in n8n, outside Workflow Ops.
 */
export const LifecycleStage = {
  INGEST: 'ingest',
  SNAPSHOT: 'snapshot',
  PROMOTION: 'promotion',
  DRIFT: 'drift',
  RECONCILIATION: 'reconciliation',
  OBSERVABILITY: 'observability',
} as const;

export type LifecycleStage = typeof LifecycleStage[keyof typeof LifecycleStage];

/**
 * Drift mode definitions for frontend.
 */
export const DriftMode = {
  PASSIVE: 'passive',
  MANAGED: 'managed',
  ENFORCED: 'enforced',
} as const;

export type DriftMode = typeof DriftMode[keyof typeof DriftMode];

/**
 * Plan-based drift mode mapping.
 */
export const PLAN_DRIFT_MODE: Record<string, DriftMode> = {
  free: DriftMode.PASSIVE,
  pro: DriftMode.PASSIVE,
  agency: DriftMode.MANAGED,
  enterprise: DriftMode.ENFORCED,
};

/**
 * Get drift mode for a plan.
 */
export function getDriftModeForPlan(planName: string): DriftMode {
  const planLower = planName?.toLowerCase() || 'free';
  return PLAN_DRIFT_MODE[planLower] || DriftMode.PASSIVE;
}

/**
 * Check if drift incidents can be created in the given drift mode.
 */
export function canCreateDriftIncident(driftMode: DriftMode): boolean {
  return driftMode === DriftMode.MANAGED || driftMode === DriftMode.ENFORCED;
}

File: n8n-ops-backend/app/core/lifecycle.py
"""
Lifecycle stage definitions for Workflow Ops.

The canonical lifecycle represents the journey of a workflow through
the Workflow Ops platform, from authoring in n8n to observability.
"""
from enum import Enum


class LifecycleStage(str, Enum):
    """
    Canonical lifecycle stages for workflows in Workflow Ops.
    
    The lifecycle flow:
    1. AUTHORING - Workflow is created/edited in n8n (outside Workflow Ops)
    2. INGEST - Workflow is synced into Workflow Ops
    3. SNAPSHOT - Workflow is committed to Git (source of truth)
    4. PROMOTION - Promotion intent is created
    5. DEPLOYMENT - Workflow is deployed to target environment
    6. DRIFT_DETECTION - System detects differences from Git source
    7. DRIFT_HANDLING - Drift is handled based on mode (passive/managed/enforced)
    8. RECONCILIATION - Drift is reconciled (promote to Git, revert, or replace)
    9. OBSERVABILITY - Workflow execution and health monitoring
    """
    AUTHORING = "authoring"
    INGEST = "ingest"
    SNAPSHOT = "snapshot"
    PROMOTION = "promotion"
    DEPLOYMENT = "deployment"
    DRIFT_DETECTION = "drift_detection"
    DRIFT_HANDLING = "drift_handling"
    RECONCILIATION = "reconciliation"
    OBSERVABILITY = "observability"

